books/kestrel/crypto/r1cs/sparse/p1cs.lisp

```diff
@@ -158,20 +162,41 @@
   (list (nums-to-strings (fp1))
         (nums-to-strings (fp2)))))
 
+(in-theory (disable (:e nums-to-strings)))
+
+(defthm alistp-of-first-nums-to-strings
+    (alistp (car (nums-to-strings)))
+  :hints (("Goal" :in-theory (enable nums-to-strings))))
+
+(defthm alistp-of-second-nums-to-strings
+    (alistp (cadr (nums-to-strings)))
+  :hints (("Goal" :in-theory (enable nums-to-strings))))
+
 ;; This shows there are no collisions:
 ;; (assert-equal (len *nums-to-strings*) (len (remove-duplicates (strip-cars *nums-to-strings*))))
 
 (defun p1cs-negative (fp x)
+  (declare (xargs :guard (and (natp fp) (< 1 fp) (integerp x))))
   (if (> x (/ fp 2))
       (str::int-to-dec-string (- x fp))
     nil))
 
+(defun well-formed-term-p (term)
+  (declare (xargs :guard t))
+  (and (true-listp term)
+       (= (len term) 2)
+       (integerp (first term))))
```

Comment:
I think you can avoid introducing a new function here and use `sparse-vector-elementp` instead. Similarly, you can use `sparse-vectorp` instead of `all-well-formed-terms-p`.